internal/services/rate.go

```diff
@@ -1,115 +1,160 @@
 package services
 
 import (
+	"context"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"net/http"
-	"sync"
+	"strings"
 	"time"
-)
 
-const RatesAPIBaseURL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
+	"github.com/fdemchenko/exchanger/internal/cache"
+	"github.com/rs/zerolog/log"
+)
 
 const (
-	RetryInterval = 500 * time.Millisecond
-	RetryCount    = 3
+	NBUExchangeRateURL        = "https://bank.gov.ua/NBUStatService/v1/statdirectory/exchange?json"
+	FawazAhmedExchangeRateURL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
+
+	DefaultCachingDuration = 15 * time.Minute
+	RateFetchTimeout       = 10 * time.Second
 )
 
-type Rate struct {
+var ErrInvalidCurrencyCode = errors.New("invalid currency code")
+
+type ExchangeRateFetcher func(code string, client *http.Client) (float32, error)
```

Comment:
There is a convention to name delegates like this `SomeActionFunc`
e.g. `FetchExchangeRateFunc`